lightningd/chaintopology.c

```diff
@@ -375,24 +376,35 @@ static u32 interp_feerate(const struct feerate_est *rates, u32 blockcount)
 
 	/* Interpolate, eg. blockcount 10, rate 15000, blockcount 20, rate 5000.
 	 * At 15, rate should be 10000.
-	 * 15000 + (15 - 10) / (20 - 10) * (5000 - 15000)
-	 * 15000 + 5 / 10 * -10000
+	 * 15000 + (15 - 10) / (20 - 10) * (15000 - 5000)
+	 * 15000 + 5 / 10 * 10000
 	 * => 10000
 	 */
+	/* Don't go backwards though! */
+	if (before->rate < after->rate)
+		return before->rate;
+
 	return before->rate
-		+ ((double)(blockcount - before->blockcount)
-		   / (after->blockcount - before->blockcount)
-		   * (after->rate - before->rate));
+		- ((u64)(blockcount - before->blockcount)
```

Comment:
Any chance for an underflow here when casting to `u64`? `blockcount` and `before->blockcount` seem not to be compared, so that invariant likely comes from outside, but an `assert` with a comment could be helpful here.

Reply:
I also assume that the `u64` is rather small, i.e., fits in a `u32` or at least the result of the calculation does, otherwise the return type could cause us to truncate the integer.

Reply:
So, blockcount >= before->blockcount, because that's the definition of "before":

```
	/* Find before and after. */
	for (size_t i = 0; i < tal_count(rates); i++) {
		if (rates[i].blockcount <= blockcount) {
			before = &rates[i];
		} else if (rates[i].blockcount > blockcount && !after) {
			after = &rates[i];
		}
	}
```

I did, however, merge this into the previous patch, since there's no reason for this fixup to be here!

The values are all 32 bits, so that's why we use 64 bits: multiplying a feerate by a block count will always fit:

```
		- ((u64)(blockcount - before->blockcount)
		   * (before->rate - after->rate)
		   / (after->blockcount - before->blockcount));
```